File: pset6/credit/credit.py
from cs50 import get_int

# credit.py

cardDB = {
    "vendor1": {"vendor": "AMEX", "qualifiers": [31, 34, 37, 5], "digits": [15]},
    "vendor2": {"vendor": "MASTERCARD", "qualifiers": [51, 52, 53, 54, 55, 13, 44, 0, 9], "digits": [16]},
    "vendor3": {"vendor": "VISA", "qualifiers": [4, 11, 81, 22], "digits": [13, 16]}
}

# get input
number = get_int("Number: ")


def calc(number):
    """ calculate the total value of alternating digits in number """
    # pass the user input to a list
    number_list = [int(i) for i in str(number)]

    num_second = ""
    num_first = 0
    count = 0

    # multiply every number in the list by 2 starting at -2
    for num in number_list[-2::-2]:
        num_second += str(num * 2)

    number_list2 = [int(i) for i in str(num_second)]
    for num3 in number_list2:
        count += num3

    # sum every 2nd number starting from -1
    for num2 in number_list[::-2]:
        num_first += num2

    return num_first + count


def cardVendor():
    """ verify the card vendor """
    card_validity = True

    # last digit as a product of Luhn’s Algorithm
    last_digit = abs(calc(number)) % 10
    # length of number
    len_number = len(str(number))
    # last two digits of number
    two_numbers = abs(number) % 100

    for key in cardDB:
        if two_numbers > 9 and two_numbers in cardDB[key]["qualifiers"]:
            if len_number in cardDB[key]["digits"] and last_digit == 0:
                print(cardDB[key]["vendor"])
                card_validity = False
                break
        elif (two_numbers % 10) in cardDB[key]["qualifiers"]:
            if len_number in cardDB[key]["digits"] and last_digit == 0:
                print(cardDB[key]["vendor"])
                card_validity = False
                break

    if card_validity:
        print("INVALID")


cardVendor()

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <stdbool.h>

// prototypes
int verifyCard(long last_two_digits, int end_digit, int digit_count);
int everyOther(int end_digit, int switch_number);
int calcSecondNumbers(int end_digit, int second_number_total);
int calcCheck(int first_number_total, int second_number_total);

int main(void)
{
    int digit_count = 0;
    int end_digit;
    long last_two_digits;
    long remaining_digits;
    int switch_number = 0;
    int first_number_total = 0;
    int second_number_total = 0;

    // ask for a card number from the user
    long card_number = get_long("Number: ");

    //find the last two digits in the number
    last_two_digits = card_number;
    while (last_two_digits >= 100)
    {
        last_two_digits /= 10;
    }

    // get the last digit in the number using a modulo operator and the
    // iterate through the number in reverse by a factor of ten each step
    while (card_number != 0)
    {
        end_digit = card_number % 10;
        switch_number = everyOther(end_digit, switch_number);
        digit_count++;

        // for every index position + 1 starting at 0 sum the output
        if (switch_number == 1)
        {
            first_number_total = first_number_total + end_digit;
        }
        else
        {
            second_number_total = calcSecondNumbers(end_digit, second_number_total);
        }

        card_number /= 10;
    }

    // check all conditions and verify the card
    if (card_number == 0 && calcCheck(first_number_total, second_number_total) == true)
    {
        verifyCard(last_two_digits, end_digit, digit_count);
    }
    else
    {
        printf("INVALID\n");
    }
}

// verify the card vendor
int verifyCard(long last_two_digits, int end_digit, int digit_count)
{
    // check for AMEX
    if (last_two_digits == 34 || last_two_digits == 37)
    {
        if (digit_count == 15)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    // check for MASTERCARD
    else if (last_two_digits == 51 || last_two_digits == 52 || last_two_digits == 53 || last_two_digits == 54 || last_two_digits == 55)
    {
        if (digit_count == 16)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else if (end_digit == 4)
    {
        // check for VISA
        if (digit_count == 13 || digit_count == 16)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
    return 0;
}

// sort the incoming numbers from our while loop by [index position]+1
int everyOther(int end_digit, int switch_number)
{
    if (switch_number == 0)
    {
        switch_number++;
    }
    else
    {
        switch_number--;
    }
    return switch_number;
}

// calculate the total of all the second numbers
int calcSecondNumbers(int end_digit, int second_number_total)
{
    int second_number = 0;
    int check_second_number;

    second_number = end_digit * 2;

    // check for numbers larger than 2 digits and sum their individual values
    if (second_number > 9)
    {
        check_second_number = second_number % 10;

        if (second_number >= 10)
        {
            second_number /= 10;
            second_number += check_second_number;
        }

    }
    return second_number_total += second_number;
}

// check if the last digit's value is 0
int calcCheck(int first_number_total, int second_number_total)
{
    bool checkvalue;
    if ((first_number_total + second_number_total) % 10 == 0)
    {
        checkvalue = true;
    }
    else
    {
        checkvalue = false;
    }
    return checkvalue;
}